src/mapping/query/arithmetic.rs

```diff
@@ -0,0 +1,406 @@
+        }
+    }
+    (left, right)
+}
+
+// Degrades non-float numerical types into floats for the purposes of convenient
+// boolean comparison.
+fn compare_number_types(
+    left: Value,
+    right: Value,
+    compare_fn: &dyn Fn(f64, f64) -> bool,
+) -> Result<Value> {
+    let (left, right) = consistent_number_types(left, right);
+    Ok(Value::Boolean(match left {
+        Value::Float(fl) => match right {
+            Value::Float(fr) => compare_fn(fl, fr),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        Value::Integer(il) => match right {
+            Value::Integer(ir) => compare_fn(il as f64, ir as f64),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        vl => return Err(format!("unable to compare left-hand field type {:?}", vl)),
+    }))
+}
+
+impl Function for Arithmetic {
+    fn execute(&self, ctx: &Event) -> Result<Value> {
+        let left = self.left.execute(ctx);
+        let right = self.right.execute(ctx);
+
+        // TODO: A lot of these comparisons could potentially be baked into the
+        // Value type. However, we would need to agree on general rules as to
+        // how different types are compared.
+
+        Ok(match self.op {
+            Operator::Multiply => {
+                let (left, right) = consistent_number_types(left?, right?);
+                match left {
+                    Value::Float(fl) => match right {
+                        Value::Float(fr) => Value::Float(fl * fr),
+                        vr => {
+                            return Err(format!(
+                                "unable to multiply right-hand field type {:?}",
+                                vr
+                            ))
+                        }
+                    },
+                    Value::Integer(il) => match right {
+                        Value::Integer(ir) => Value::Integer(il * ir),
+                        vr => {
+                            return Err(format!(
+                                "unable to multiply right-hand field type {:?}",
+                                vr
+                            ))
+                        }
+                    },
+                    vl => return Err(format!("unable to multiply left-hand field type {:?}", vl)),
+                }
+            }
+
+            Operator::Divide => {
+                let (left, right) = consistent_number_types(left?, right?);
+                match left {
+                    Value::Float(fl) => match right {
+                        Value::Float(fr) => Value::Float(fl / fr),
+                        vr => {
+                            return Err(format!("unable to divide right-hand field type {:?}", vr))
+                        }
+                    },
+                    Value::Integer(il) => match right {
+                        Value::Integer(ir) => Value::Float(il as f64 / ir as f64),
```

Comment:
If you are converting the result of integer division into a float, we should have a separate plain integer division operation. Python 3 went this way with `/` and `//`.

Reply:
See: https://github.com/timberio/vector/issues/3729